### PR TITLE
Parallel loading for Cypher

### PR DESCRIPTION
**Pull request by jexp**

### CODE REVIEW

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java [hunk 1]
```diff
@@ -88,8 +260,48 @@ public boolean visit(Result.ResultRow row) throws RuntimeException {
...
+        final WeightMapping nodeProps = newWeightMapping(hasNodeProperty, setup.nodeDefaultPropertyValue, capacity);
+
+        class NodeRowVisitor implements Result.ResultVisitor<RuntimeException> {
+            private long rows;
+            @Override
+            public boolean visit(Result.ResultRow row) throws RuntimeException {
+                rows++;
+                long id = row.getNumber("id").longValue();
+                idMap.add(id);
+                if (hasNodeWeights) nodeWeights.set(id, row.get("weight"));
```
Comment (knutwalker) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java]: Can we have all the expected row variable names as constants somewhere? Same for `source`, `target`, `value` and `id`?

File: core/src/main/java/org/neo4j/graphalgo/api/GraphSetup.java [hunk 2]
```diff
@@ -120,11 +130,19 @@ public GraphSetup(ExecutorService executor) {
...
         this.relationshipStatement = null;
+        this.batchSize = -1;
+        this.accumulateWeights = false;
     }
 
     public boolean loadConcurrent() {
         return executor != null;
     }
+    public int concurrency() {
+        if (!loadConcurrent()) return 1;
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/api/GraphSetup.java]: Braces

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java [hunk 3]
```diff
@@ -110,25 +111,53 @@ public void growIn(int targetNodeId, int length) {
...
+        }
         outgoing[sourceNodeId][degree] = targetNodeId;
         outgoingIds[sourceNodeId][degree] = relationId;
         outOffsets[sourceNodeId] = nextDegree;
     }
 
     /**
+     * checks for outgoing target node
+     */
+    public boolean hasOutgoing(int sourceNodeId, int targetNodeId) {
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java]: O(n) for an existance-check

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java [hunk 4]
```diff
@@ -110,25 +111,53 @@ public void growIn(int targetNodeId, int length) {
...
+        }
         incoming[targetNodeId][degree] = sourceNodeId;
         incomingIds[targetNodeId][degree] = relationId;
         inOffsets[targetNodeId] = nextDegree;
     }
 
     /**
+     * checks for incoming target node
+     */
+    public boolean hasIncoming(int sourceNodeId, int targetNodeId) {
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java]: O(n)

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java [hunk 5]
```diff
@@ -258,6 +287,22 @@ private void forEachIncoming(int nodeId, WeightMapping weights, WeightedRelation
         }
     }
 
+    public void forEachNode(Direction direction, IntPredicate nodeConsumer) {
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java]: That functionality should be implemented by a NodeIterator

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java [hunk 6]
```diff
@@ -1,71 +1,243 @@
...
+                    working = result.rows > 0;
+                    if (working) {
+                        WeightMapping resultWeights = hasRelationshipWeights && result.relWeigths.size() > 0 ? result.relWeigths : null;
+                        result.matrix.forEachNode(Direction.OUTGOING,
+                                node -> {
+                                    result.matrix.forEach(node, Direction.OUTGOING,
+                                            (source, target, relationship) -> {
+                                                if (accumulateWeights) {
+                                                    // suboptimial, O(n) per node
+                                                    if (!matrix.hasOutgoing(source, target)) {
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java]: This wont work on big dense graphs

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java (hunk 1 above)
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java]: braces

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java [hunk 7]
```diff
@@ -88,8 +260,48 @@ public boolean visit(Result.ResultRow row) throws RuntimeException {
...
+
+        class NodeRowVisitor implements Result.ResultVisitor<RuntimeException> {
+            private long rows;
+            @Override
+            public boolean visit(Result.ResultRow row) throws RuntimeException {
+                rows++;
+                long id = row.getNumber("id").longValue();
+                idMap.add(id);
+                if (hasNodeWeights) nodeWeights.set(id, row.get("weight"));
+                if (hasNodeProperty) nodeProps.set(id, row.get("value"));
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java]: braces

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java [hunk 8]
```diff
@@ -88,8 +260,48 @@ public boolean visit(Result.ResultRow row) throws RuntimeException {
...
+
+    private WeightMapping newWeightMapping(boolean needWeights, double defaultValue, int capacity) {
+        return needWeights ?
+                new WeightMap(capacity, defaultValue) :
+                new NullWeightMap(defaultValue);
+    }
 
-        return new HeavyGraph(idMap, matrix, relWeigths, nodeWeights, nodeProps);
+    private Map<String, Object> params(long offset, int batchSize) {
+        if (batchSize > 0) return MapUtil.map(SKIP,offset, LIMIT,batchSize);
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java]: braces

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java [hunk 9]
```diff
@@ -110,25 +111,53 @@ public void growIn(int targetNodeId, int length) {
...
     }
 
     /**
+     * checks for incoming target node
+     */
+    public boolean hasIncoming(int sourceNodeId, int targetNodeId) {
+        final int degree = inOffsets[sourceNodeId];
+        int[] rels = incoming[sourceNodeId];
+        for (int offset = degree-1; offset >= 0; offset--) {
+            if (rels[offset]==targetNodeId) return true;
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java]: braces pls

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java [hunk 10]
```diff
@@ -1,71 +1,243 @@
...
     }
 
+    static class Relationships {
+
+        private final long offset;
+        private final long rows;
+        private final AdjacencyMatrix matrix;
+        private final WeightMapping relWeigths;
+
+        Relationships(long offset, long rows, AdjacencyMatrix matrix, WeightMapping relWeigths) {
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java]: typo

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java [hunk 11]
```diff
@@ -110,25 +111,53 @@ public void growIn(int targetNodeId, int length) {
...
     }
 
     /**
+     * checks for outgoing target node
+     */
+    public boolean hasOutgoing(int sourceNodeId, int targetNodeId) {
+        final int degree = outOffsets[sourceNodeId];
+        int[] rels = outgoing[sourceNodeId];
+        for (int offset = degree-1; offset >= 0; offset--) {
+            if (rels[offset]==targetNodeId) return true;
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java]: braces

File: core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java [hunk 12]
```diff
@@ -1,71 +1,243 @@
...
+    private Relationships batchLoadRelationships(int batchSize, Nodes nodes) {
+        ExecutorService pool = setup.executor;
+        int threads = setup.concurrency();
+        boolean accumulateWeights = setup.accumulateWeights;
+
+        // data structures for merged information
+        int nodeCount = nodes.idMap.size();
+        AdjacencyMatrix matrix = new AdjacencyMatrix(nodeCount);
         boolean hasRelationshipWeights = !setup.loadDefaultRelationshipWeight();
+        final WeightMapping relWeigths = newWeightMapping(hasRelationshipWeights, setup.relationDefaultWeight, nodeCount*ESTIMATED_DEGREE);
```
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java]: typo

**Pull request merged**